client/src/main/java/io/pravega/client/netty/impl/SessionHandler.java

```diff
@@ -0,0 +1,235 @@
+        // WireCommands.KeepAlive messages are sent for every network connection to a SegmentStore.
+        ScheduledFuture<?> old = keepAliveFuture.getAndSet(ch.eventLoop().scheduleWithFixedDelay(new KeepAliveTask(), 20, 10, TimeUnit.SECONDS));
+        if (old != null) {
+            old.cancel(false);
+        }
+    }
+
+    /**
+     * Invoke all the {@link ReplyProcessor#connectionDropped()} for all the registered sessions once the
+     * connection is disconnected.
+     *
+     * @see io.netty.channel.ChannelInboundHandler#channelUnregistered(ChannelHandlerContext)
+     */
+    @Override
+    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+        registeredFutureLatch.reset();
+        ScheduledFuture<?> future = keepAliveFuture.get();
+        if (future != null) {
+            future.cancel(false);
+        }
+        channel.set(null);
+        sessionIdReplyProcessorMap.forEach((sessionId, rp) -> {
+            rp.connectionDropped();
+            log.debug("Connection dropped for session id : {}", sessionId);
+        });
+        super.channelUnregistered(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        Reply cmd = (Reply) msg;
+        log.debug(connectionName + " processing reply: {} with session {}.", cmd, Session.from(cmd.getRequestId()));
+
+        if (cmd instanceof WireCommands.Hello) {
+            sessionIdReplyProcessorMap.forEach((sessionId, rp) -> rp.hello((WireCommands.Hello) cmd));
+            return;
+        }
+
+        if (cmd instanceof WireCommands.DataAppended) {
+            batchSizeTracker.recordAck(((WireCommands.DataAppended) cmd).getEventNumber());
+        }
+        // Obtain ReplyProcessor and process the reply.
+        getReplyProcessor(cmd).ifPresent(processor -> {
+            try {
+                processor.process(cmd);
+            } catch (Exception e) {
+                processor.processingFailure(e);
+            }
+        });
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+        sessionIdReplyProcessorMap.forEach((sessionId, rp) -> {
+            rp.processingFailure(new ConnectionFailedException(cause));
+            log.debug("Exception observed for session id : {}", sessionId);
+        });
+    }
+
+    @Override
+    public void close() {
+        if (closed.compareAndSet(false, true)) {
+            Channel ch = channel.get();
+            if (ch != null) {
+                log.debug("Closing channel:{} ", ch);
+                log.warn("{} sessions are not closed", sessionIdReplyProcessorMap.size());
+                ch.close();
+            }
+        }
+    }
+
+    private final class KeepAliveTask implements Runnable {
+        @Override
+        public void run() {
+            try {
+                if (!recentMessage.getAndSet(false)) {
```

Comment:
Why do you set to `false` and then immediately you set it to `true`?

Reply:
Fixed it.